### PR TITLE
feat(common): sports_helpers — the helpers all nine scoreboards carry identical copies of

### PR DESCRIPTION
**Pull request by ChuckBuilds**
## What

A new hardware-free module, `src/common/sports_helpers.py`, holding the helpers the scoreboard plugins' `sports.py` carry **byte-identical** copies of (docstring-stripped AST) and that have no equivalent elsewhere in `src/common`. Bodies were copied from the plugin copies, not from `src/base_classes`, at **ledmatrix-plugins `f09bff2`** (origin/main, after #491 and #492).

Phase A1 of the base_classes → `src/common` plan. Nothing in core imports it yet: **no behaviour change**.

## Why a new module, not more methods on `sports_shared.py`

Suppose a plugin deletes its copy of a method and relies on an *existing* module having grown it. On an older core that fails at runtime with `AttributeError` mid-render, usually swallowed into a blank board. Neither the plugin loader nor the plugins' `scripts/check_min_core_version.py` (which checks modules) can see that. A module that doesn't exist yet fails at load instead, where both checks can.

## Promoted

| Here | Plugin name | Identical in |
|---|---|---|
| `clamp_window` (+ `MIN_WINDOW_DAYS`, `MAX_WINDOW_DAYS`) | `_clamp_window`, `_MIN_/_MAX_WINDOW_DAYS` | all 9 |
| `clamp_seconds` | `_clamp_seconds` | all 9 |
| `logo_needs_refresh` (lazy `src.logo_downloader` import, as in the plugins) | `_logo_needs_refresh` | all 9 |
| `spread_weighted_order` / `SportsHelpersMixin._spread_weighted_order` | `SportsCore._spread_weighted_order` | all 9 |
| `SportsHelpersMixin._mode_customization` | same | all 9 |
| `SportsHelpersMixin._setting_int` | same | all 9 |
| `SportsHelpersMixin._reset_dwell_on_reentry` (+ `_DWELL_REENTRY_GAP_SECONDS = 5.0`) | same | all 9 |
| `SportsHelpersMixin._next_switch_index` | same | all 9 |
| `SportsHelpersMixin._odds_color` | same | 8 (ufc has none) |
| `SportsHelpersMixin._upcoming_date_and_time_text` | same | 8 (ufc has none) |

"All 9" = afl, baseball, basketball, football, hockey, lacrosse, nrl, soccer, ufc. Mixin methods keep the plugins' exact names and signatures, so adopting it means deleting the copies and adding one base.

Also `_favorite_key(game, side)`, the override point from `src/base_classes/sports/core.py` (default `game.get(f"{side}_abbr")`). No plugin defines it and nothing here calls it. It's carried so later phases (celebrations, game selection) have a hardware-free home for the seam.

**Excluded:** none of the ten candidates. All were identical at `f09bff2`.

**Found but not added** (identical in ≥8 `SportsCore` copies, per the same scan):
- `_rotate_other_games_on_display`: identical in the 8, ufc differs, no `src/common` equivalent. A candidate for a later phase. Not added here because it wasn't in the planned set and needs its own hardware and host-dependency review.
- `_get_timezone`, `_extract_game_details`, `_fetch_data`: identical in 9 but deliberately per-plugin. The first binds a per-plugin timezone module; the other two are the abstract sport contract. See the `sports_shared.py` docstring.
- `_FONT_NAME_ALIASES`, `_FONT_PIXEL_GRID`, `_NARROW_SCORE_RUNGS`: already on `SportsCoreSharedMixin`.

## Host contract (full list in the module docstring, enforced by a test)

- `config`: `_mode_customization`
- `mode_config`, `logger` (and optionally `league`): `_setting_int`
- `games_list`, `current_game_index`, and `_is_favorite_game` (only when boost > 1): `_next_switch_index`
- `last_game_switch`: `_reset_dwell_on_reentry`
- `_card_option`, `_format_game_date`, `_format_game_time`: `_upcoming_date_and_time_text`, all from `SportsCoreSharedMixin`
- `_element_color` (optional, falls back to green): `_odds_color`
- `SKIN_MODE` (optional)
- Class defaults: `_DWELL_REENTRY_GAP_SECONDS = 5.0`, `favorite_rotation_boost = 1`

No `__init__` and no bare `super()`, so base order doesn't matter. By convention it goes after `SportsCoreSharedMixin`.

## Tests

- `test/test_sports_helpers.py`
  - Behaviour tests for every helper, ported from the plugins' tests: clamps including ±Infinity/`OverflowError`, placeholder refresh including an older core without the markers, dwell re-entry, `spread_weighted_order` ordering and wrap, `_next_switch_index` with `favorite_rotation_boost`, `_setting_int`, `_mode_customization`, `_odds_color`, and the `switch_show_date`/`switch_show_time` toggles.
  - A host-contract test that derives every `self.X` read and requires it in the docstring, plus a minimal `SportsCoreSharedMixin + SportsHelpersMixin` host in both base orders.
  - A **parity test**: with `LEDMATRIX_PLUGINS` pointing at a plugins checkout, every promoted body is AST-compared against every plugin copy that carries it, along with the constants. It skips without the variable, since CI has no plugins checkout.
- `test/test_common_is_hardware_free.py`
  - (a) In a fresh interpreter with `sys.modules['rgbmatrix'] = None`, imports `src.common` and every `src.common.sports_*`, and asserts no `src.display_manager` / `src.base_classes` module got loaded.
  - (b) An AST scan failing on any module-level import (including inside `try`/`if`) of `src.base_classes`, `src.display_manager` or `src.plugin_system` in `src/common/*.py`. **No existing module violated it**, so the allowlist is empty; a stale-entry check keeps it that way.

## Docs

- `CHANGELOG.md`: `## Unreleased` entry in the 3.3.0 module-addition style. `__version__` stays 3.4.0; the next release decides 3.5.0.
- `docs/SPORTS_UNIFICATION.md`: the module is added to the layering block, with a short "Converging on `src/common`" note that points at the plugins repo's `08-shared-sports-code.md` for adoption steps.

## Follow-up

- Plugins adopt once a release ships this and they floor `ledmatrix_min_version` on it (plan phase C1: the 8 lineage-sharing scoreboards import `sports_helpers` and delete their copies; ufc can adopt the free functions and the 5 methods it carries).
- Deprecating `base_classes` is a later PR.

## Verification

- **Full suite** (Windows, Python 3.14), branch vs pristine origin/main `814c21de`: 114 failed / 4921 passed / 93 skipped vs 114 failed / 4851 passed / 74 skipped. **The failing set is identical, so no new failures.** The 114 are the known Windows-only failures. The +19 skips are the parity tests without `LEDMATRIX_PLUGINS`.
- **Parity test run for real** with `LEDMATRIX_PLUGINS` pointed at a detached worktree of plugins `f09bff2`: all 19 parity cases passed (10 bodies × every carrier, plus constants for all 9). Sanity check: changing one literal in a copy of hockey's `_odds_color` fails the test and names hockey.
- New tests plus `test_version_consistency.py` and `test_sports_shared.py`: all pass.
- `python scripts/check_release_version.py v3.4.0`: OK (`__version__` unchanged).
- Hardware-free: `src.common` and all `src.common.sports_*` import with `rgbmatrix` blocked, and load no `src.display_manager` / `src.base_classes` / `src.plugin_system`. No existing `src/common` module imports those at module level.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added shared sports scoreboard helpers for plugin authors, including settings validation, game rotation, favorites handling, logo refresh checks, odds colors, and date/time formatting.
  - The shared module can be used without display hardware, improving compatibility in supported environments.

- **Documentation**
  - Added guidance for adopting the shared sports helpers, maintaining plugin compatibility, and migrating existing bundled helper code.
  - Updated the unreleased changelog with plugin usage and compatibility notes.

- **Tests**
  - Added coverage for sports helper behavior, edge cases, hardware-free imports, and compatibility across scoreboard plugins.

### CODE REVIEW

File: src/common/sports_helpers.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    for index, weight in enumerate(weights):
+        for turn in range(weight):
+            slots.append(((index + turn * count / weight) % count, turn > 0, index))
+    return [index for _, _, index in sorted(slots)]
+
+
+class SportsHelpersMixin:
+    """The ``SportsCore`` methods identical across the scoreboards' sports.py.
+
+    Stateless and constructor-free; see the module docstring for the host
```
[the remark this answers is not in the document]
Comment (ChuckBuilds) [reply src/common/sports_helpers.py]: Reworded in 4d86b98f: the `SportsHelpersMixin` docstring now says it is constructor-free but keeps lazy state on its host (see the module docstring), and the CHANGELOG entry says the same instead of "No `__init__` and no state".

File: test/test_common_is_hardware_free.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    modules = sorted(p.stem for p in COMMON.glob("sports_*.py"))
+    assert "sports_helpers" in modules
+    script = textwrap.dedent(f"""
+        import importlib, json, sys
+        sys.modules["rgbmatrix"] = None   # any import of it now raises
+        importlib.import_module("src.common")
+        for name in {modules!r}:
+            importlib.import_module("src.common." + name)
+        loaded = sorted(m for m in sys.modules
+                        if m.startswith(("src.display_manager", "src.base_classes")))
```
[the remark this answers is not in the document]
Comment (ChuckBuilds) [reply test/test_common_is_hardware_free.py]: Fixed in 4d86b98f. The runtime check now filters every `FORBIDDEN` package (so `src.plugin_system` too); nothing loads it today, so the test stays green. `_targets()` resolves relative imports against the scanned module's package (`src.common`), so `from .. import plugin_system` and `from ..plugin_system import x` are caught. New guard tests cover both forms, directly and end to end through `_violations()`, and the absolute cases are kept.

File: test/test_sports_helpers.py [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+
+    @pytest.mark.parametrize("sport", SCOREBOARDS)
+    def test_constants_match(self, sport):
+        root = _plugins_root()
+        defs = _definitions(ast.parse(
+            (root / f"{sport}-scoreboard" / "sports.py").read_text(encoding="utf-8")))
+        assert ast.literal_eval(defs["module"]["_MIN_WINDOW_DAYS"].value) == MIN_WINDOW_DAYS
+        assert ast.literal_eval(defs["module"]["_MAX_WINDOW_DAYS"].value) == MAX_WINDOW_DAYS
+        gap = defs["SportsCore"]["_DWELL_REENTRY_GAP_SECONDS"].value
+        assert math.isclose(ast.literal_eval(gap), SportsHelpersMixin._DWELL_REENTRY_GAP_SECONDS)
```
[the remark this answers is not in the document]
Comment (ChuckBuilds) [reply test/test_sports_helpers.py:603]: Agreed that the skip is not coverage. The CI side is handled in plugins: ChuckBuilds/ledmatrix-plugins#495 runs the same comparison in plugins CI (`scripts/check_sports_helpers_parity.py`), which checks out core. The core CI config is unchanged. As of 4d86b98f the skip reason and the module docstring name that guard, so a skip here is not read as coverage. Locally, with `LEDMATRIX_PLUGINS` set to plugins main, all parity cases pass.
